README.txt
Zebedee Secure Tunnel
=====================

Zebedee is a simple program to establish an encrypted, compressed
"tunnel" for TCP/IP or UDP traffic between two systems. This
allows data from, for example, telnet, ftp and X sessions to be
protected from snooping. You can also use compression, either
with or without data encryption, to gain performance over
low-bandwidth networks.

The main goals for Zebedee are to:

 * Provide client and server functionality under both UNIX/Linux
   and Windows.

 * Be easy to install, use and maintain with little or no
   configuration required.

 * Have a small footprint, low wire protocol overhead and
   give significant traffic reduction by the use of
   compression.

 * Use only algorithms that are either unpatented or for
   which the patent has expired.

 * Be entirely free for commercial or non-commercial use and
   distributed under the term of the GNU General Public
   Licence (see LICENCE.txt for details).

For further information on how to use Zebedee see the file
zebedee.html in the distribution (or the manual page for
zebedee(1) under UNIX -- it is basically the same text). Example
configuration files are also provided.

Building Zebedee
----------------

For instructions on how to build Zebedee see the file
BUILDING.txt in the distribution.

Special Notes for Windows Installations
---------------------------------------

The installation on Windows systems creates a Start Menu entry
that contains an icon to start a Zebedee server. If you select
this menu entry Zebedee will take its configuration information
from the file "server.zbd" in the Zebedee installation directory.
You MUST edit this file before you can do anything useful with
Zebedee. The "out of the box" configuration is deliberately
as secure as possible and will not allow you create any tunnels
to the server without you explicitly enabling them.

The Zebedee service is not installed automatically. If you want
to run Zebedee as a service you must use the "-Sinstall" option
from the command-line.

Zebedee clients will probably, in general, be started from the
command line. However, a file-type association is created by
the installation for files of type ".zbd". This can be used to
launch Zebedee clients automatically from Explorer or elsewhere
-- provided that all the necessary information is contained
within the configuration file.

Roman Hoog Antink, 28.4.2011


CHANGES.txt
* Release 2.5.7

Add -a option to set outgoing source IP.

* Release 2.5.6

Fix address size for Solaris IPv6 bind() and sendto() calls.

Update --help output, adding option -4.

* Release 2.5.5

Fix compilation of IPv6 support on Mac OS X.

Add command line option -4 to enforce local IPv4 sockets when compiled with
IPv6 support.

Minor IPv6 fix.

Makefile comments completion (system libs).

* Release 2.5.4

Add IPv6 support.

Update to latest libraries (zlib, bzip2 and blowfish/OpenSSL).

Minor fixes (typos and compiler warnings).

* Release 2.5.3

Fix file descriptor leak when IP address lookup fails.

Fix problem with running a server in "reverse" mode and detached -- only
apparent on Windows.

Add "maxconnections" to alleviate DoS attack.

Check for target port 0 to avoid DoS.

Linux 64-bit port (a result of the "Linux on POWER" contest) courtesy of
Stew Benedict <stewb@linuxcontrol.net>. Use the "linux64" OS target.

Upgraded version of bzip2 and zlib.

* Release 2.5.2

Cross-platform fix for checksumming code. This is *incompatible* with version
2.5.1. As a temporary workaround, setting "bugcompatibility 251" will maintain
compatibility with release 2.5.1 for little-endian platforms (e.g. Intel).
This will be removed from the final production release.

Upgrade to Inno Setup 4.

More documentation fixes.

Increased the default thread stack size to 64k and added "threadstacksize" for
debug/test purposes.

Fix handling of HTTP/1.1 responses from proxies.

Added "acceptconnecttimeout" (supersedes "connecttimeout") along with
"connectattempts" (and -C command-line option), "serverconnecttimeout" and
"targetconnecttimeout".

Fixed bug with "clienthost" not being honoured when Zebedee was used as
a service.

Added display of NOTICE.txt to the Windows install process.

* Release 2.5.1 (2003/07/05)

Critical fix to the checksumming code, which wouldn't work if messages
got out of sync between client and server. This is *incompatible* with
version 2.5.0.

Fixed documentation comments on message integrity.

* Release 2.5.0 (2003/07/02)

Dropped backwards compatibility with Zebedee 1.x versions of the protocol.

Added message integrity checking ("checksumlevel" and "minchecksumlevel"
keywords), "dropunknownprotocol" and "lockprotocol" features and per-target
address checking. These are all derived from patches submitted by Henrick
Lund. Great work Henrick!

Added the ability to hand the peer address, target address and target
port to key generation commands (add a "+" to the end of the command
to get three extra arguments provided).

Added "httpproxyauth" -- thanks to James CE Johnson.

Added "runasuser/-N" -- thanks to Thomas Melzer.

Added per-target ID checking -- originally by Alain Turbide.

Fixed a bug with "anonymous" tunnel specs (using a "*" to default to
the server address).

Fixed bug writing to syslog when detached -- thanks to Kyle Dent.

Set FD_SETSIZE to 512 by default on Windows, to allow many more simultaneous
connections to be handled correctly (approx 250). This can be overridden
at compile time. Thanks to John W Campbell.

Bug fix for 64-bit HPUX support. Thanks to Kris Laporte.

* Release 2.4.1 (2002/05/29)

Fixed incorrect selection of client TCP/UDP mode when handler function is
run in-line (as a result of debug, multiuse or command setting).

Fixed message when client requests incompatible TCP/UDP mode.

Added "dumpdata" keyword to enable a dump of local traffic when this is
true (and the verbosity level is 5 or greater).

Added support for BSD/OS (OS=bsdi) in the Makefile -- thanks to Brett Wynkoop.

* Release 2.4.0 (2002/05/09)

Added support for building on MacOS X (sorry, no support for "classic" MacOS
yet). This required a larger initial thread stack size, which can now be
specified at compile time.

Fix handling of targets containing sub-net specifications.

Eliminate use of non-reentrant inet_ntoa() calls. This has never been
observed as a problem but could, potentially, cause issues with confusing
messages or even incorrect connections.

"Modernised" vncserver.zbd configuration file.

Improved the Windows key generation code. Note, however, that Zebedee's key
generation for all platforms is still not guaranteed to be cryptographically
strong. If you are really, really, concerned with the quality of key
generation please examine the code, improve it if necessary or call out to an
external command (such as "gpg --gen-random"). However, unless you are trying
to defend very sensitive traffic against snooping from determined and
well-equipped attackers I wouldn't worry too much. And if you are trying to do
that, Zebedee may not be the right choice for you anyway :-)

Changed server tunnel establishment message to be more informative.

Fixed (implemented!) "detached" mode for server-initiated connections. It
now detaches after having contacted the client but before establishing any
tunnels.

Added -F option to specify configuration file field separator character.

Added "sharedkey" and "sharedkeygencommand" to allow a pre-established shared
secret key to be specified. This and the -F option have been added primarily
for use in Ed Suominen's PRIVARIA.

Included "passphrase.tcl" in the distribution.

Numerous changes to the (English) documentation.

* Release 2.3.2 (2002/04/18)

Tunnel and target ports can now be suffixed with "/tcp" or "/udp" in order to
restrict the protocol used. This is useful when "ipmode both" is in effect.
So, for example a target of "localhost:5900-5910/tcp,daytime/udp" allows
TCP-only tunnels to 5900-5910 and UDP-only tunnels to the "daytime" port. When
the protocol is omitted either may be used.

Added support for UDP source address spoofing when run in "transparent" mode.
This makes source to target UDP traffic "transparent" but the target server
will be unlikely to be able to reply to any such datagrams received. Note that
this is not compiled in by default and requires the use of the "libnet"
library.


service.c
/*
** This file is part of "zebedee"
**
** Copyright 1999 by Neil Winton. All rights reserved.
** 
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License as published by
** the Free Software Foundation; either version 2 of the License, or
** (at your option) any later version.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
**
** $Id: service.c,v 1.1.1.1 2001-04-12 18:07:12 ndwinton Exp $
**
** $Log: not supported by cvs2svn $
** Revision 1.2  2000/01/08 21:35:22  nwinton
** Version 1.3.0
**
** Revision 1.1  1999/09/24 20:20:58  nwinton
** Cleanup of compiler warnings
**
** Revision 1.0  1999/09/18 21:35:06  nwinton
** Initial revision
**
*/

#include <windows.h>
#include <stdio.h>
#include <stdlib.h>
#include <process.h>

char *service_c_rcsid = "$Id: service.c,v 1.1.1.1 2001-04-12 18:07:12 ndwinton Exp $";

extern void message(unsigned short, int, char *, ...);

#define MAX_BUF_SIZE    2048

HANDLE                  SvcFinishEvent = NULL;
SERVICE_STATUS          SvcStatus;      /* Current status of service */
SERVICE_STATUS_HANDLE   SvcStatusHandle;
DWORD                   SvcError;
HANDLE                  SvcThreadHandle = NULL;
char                    *SvcName = NULL;
VOID                    (*SvcFunction)(VOID *) = NULL;
VOID                    *SvcArg = NULL;

DWORD   svcPlatform(void);
VOID    svcRun(char *name, VOID (*function)(VOID *), VOID *arg);
VOID    svcRun9X(char *name);
VOID    svcRunNT(char *name);
VOID    svcMain(DWORD argc, LPTSTR *argv);
VOID    svcControl(DWORD ctrlCode);
BOOL    svcReport(DWORD currentState, DWORD exitCode, DWORD checkPoint,
                  DWORD waitHint);
VOID    svcStop(LPTSTR msg);
int     svcInstall(char *name, char *configFile);
int     svcInstall9X(char *name, char *configFile);
int     svcInstallNT(char *name, char *configFile);
int     svcRemove(char *name);
int     svcRemove9X(char *name);
int     svcRemoveNT(char *name);


DWORD
svcPlatform(void)
{
    OSVERSIONINFO info;

    info.dwOSVersionInfoSize = sizeof(info);
    if (!GetVersionEx(&info))
    {
        message(0, 0, "can't get OS version info");
        return 0;
    }
    return info.dwPlatformId;
}

VOID
svcRun(char *name, VOID (*function)(VOID *), VOID *arg)
{
    DWORD platform = svcPlatform();

    SvcName = name;
    SvcFunction = function;
    SvcArg = arg;

    switch (platform)
    {
    case VER_PLATFORM_WIN32_WINDOWS:
        svcRun9X(name);
        break;

    case VER_PLATFORM_WIN32_NT:
        svcRunNT(name);
        break;

    default:
        message(0, 0, "unsupported OS platform (type %d)", platform);
        break;
    }
}

VOID
svcRunNT(char *name)
{
    SERVICE_TABLE_ENTRY dispatchTable[] = {
        { TEXT(name), (LPSERVICE_MAIN_FUNCTION)svcMain },
        { NULL, NULL }
    };

    if (!StartServiceCtrlDispatcher(dispatchTable))
    {
        svcStop("StartServiceCtrlDispatcher failed");
    }
}

VOID
svcRun9X(char *name)
{
    HINSTANCE kernelDll;
    DWORD (*regFn)(void *, DWORD);


    /* Obtain a handle to the kernel library */

    if ((kernelDll = LoadLibrary("KERNEL32.DLL")) == NULL)
    {
        message(0, 0, "can't get handle to kernel library");
        return;
    }

    /* Find the address of the RegisterServiceProcess function */

    regFn = (DWORD (*)(void *, DWORD))GetProcAddress(kernelDll, "RegisterServiceProcess");
    if (regFn == NULL)
    {
        message(0, 0, "can't get the address of RegisterServiceProcess()");
        return;
    }
                        
    /* Register this process with the OS as a service */

    if ((*regFn)(NULL, 1 /* RSP_SIMPLE_SERVICE */) == 0)
    {
        message(0, 0, "failed to register service process");
        exit(EXIT_FAILURE);
    }

    /* Run the main routine */

    (*SvcFunction)(SvcArg);

    /* Unregister the process */

    if ((*regFn)(NULL, 0 /* RSP_UNREGISTER_SERVICE */) == 0)
    {
        message(0, 0, "failed to unregister service process");
        exit(EXIT_FAILURE);
    }

    /* Free the kernel library */

    FreeLibrary(kernelDll);

    exit(EXIT_SUCCESS);
}

/*
**  svcMain
**
** This function starts the service worker thread proper. It then waits
** for the worker to terminate.
**
*/

VOID
svcMain(DWORD argc, LPTSTR *argv)
{
    DWORD   wait;


    /* Register the control handler */

    SvcStatusHandle = RegisterServiceCtrlHandler(TEXT(SvcName),
                                                 (LPHANDLER_FUNCTION)svcControl);
    if (!SvcStatusHandle)
    {
        message(0, 0, "failed to register service control handler");
        goto finish;
    }

    /* Initialise static service status values */

    SvcStatus.dwServiceType = SERVICE_WIN32_OWN_PROCESS;
    SvcStatus.dwServiceSpecificExitCode = 0;

    /* Report to SCM */

    if (!svcReport(SERVICE_START_PENDING,   /* Service state */
                   NO_ERROR,                /* Exit code */
                   1,                       /* Checkpoint number */
                   5000))                   /* Wait hint -- 5 secs */
    {
        message(0, 0, "failed to report status to SCM");
        goto finish;
    }

    /*
    ** Create the "stop event" object. The arguments indicate a manually
    ** reset event whose initial state is unset.
    */

    if ((SvcFinishEvent = CreateEvent(NULL, 1, 0, NULL)) == NULL)
    {
        goto finish;
    }

    /* Report next checkpoint to SCM */

    if (!svcReport(SERVICE_START_PENDING, NO_ERROR, 2, 5000))
    {
        message(0, 0, "failed to report status to SCM");
        goto finish;
    }

    /* Start the worker thread */

    SvcThreadHandle = (HANDLE)_beginthread(SvcFunction, 65536, SvcArg);
    if (!SvcThreadHandle)
    {
        message(0, 0, "failed to create worker thread");
        goto finish;
    }

    /* Report all systems GO! */

    if (!svcReport(SERVICE_RUNNING, NO_ERROR, 0, 0))
    {
        message(0, 0, "failed to report status to SCM");
        goto finish;
    }

    /* Wait until SvcFinishEvent is set */

    wait = WaitForSingleObject(SvcFinishEvent, INFINITE);

finish:
    if (SvcFinishEvent != NULL)
    {
        CloseHandle(SvcFinishEvent);
    }

    /* Report to SCM if possible */

    if (SvcStatusHandle != 0)
    {
        svcReport(SERVICE_STOPPED, SvcError, 0, 0);
    }

    return;
}

/*
** svcControl
**
** Handle SCM service control requests
*/

VOID
svcControl(DWORD code)
{
    DWORD  state = SERVICE_RUNNING;
 

    switch(code)
    {
    case SERVICE_CONTROL_PAUSE:
        if (SvcStatus.dwCurrentState == SERVICE_RUNNING)
        {
            SuspendThread(SvcThreadHandle);
            state = SERVICE_PAUSED;
        }
        break;

    case SERVICE_CONTROL_CONTINUE:
        if (SvcStatus.dwCurrentState == SERVICE_PAUSED)
        {
            ResumeThread(SvcThreadHandle);
            state = SERVICE_RUNNING;
        }
        break;

    case SERVICE_CONTROL_STOP:
        state = SERVICE_STOP_PENDING;
        svcReport(SERVICE_STOP_PENDING, NO_ERROR, 1, 5000);
        SetEvent(SvcFinishEvent);
        return;

    case SERVICE_CONTROL_INTERROGATE:
        break;

    default:
        break;
    }

    svcReport(state, NO_ERROR, 0, 0);
}

/*
** svcReport
**
** Update the SCM with the current state of the service.
*/

BOOL
svcReport(DWORD currentState, DWORD exitCode, DWORD checkPoint, DWORD waitHint)
{
    BOOL result;

    /* Disable control requests while starting */

    if (currentState == SERVICE_START_PENDING)
    {
        SvcStatus.dwControlsAccepted = 0;
    }
    else
    {
        SvcStatus.dwControlsAccepted = SERVICE_ACCEPT_STOP |
                                       SERVICE_ACCEPT_PAUSE_CONTINUE;
    }

    SvcStatus.dwCurrentState = currentState;
    SvcStatus.dwWin32ExitCode = exitCode;
    SvcStatus.dwCheckPoint = checkPoint;
    SvcStatus.dwWaitHint = waitHint;

    /* Make the report */

    if (!(result = SetServiceStatus(SvcStatusHandle, &SvcStatus)))
    {
        /* If there is an error then bail out */
        svcStop("SetServiceStatus failed!");
    }

    return result;
}

/*
** svcStop
**
** Report an error and stop the service
*/

VOID
svcStop(LPTSTR msg)
{
    SvcError = GetLastError();

    message(0, 0, "failed in Win32 service routines: code %#0x", SvcError);

    /* Signal the main routine to finish */

    SetEvent(SvcFinishEvent);
}

int
svcInstall(char *name, char *configFile)
{
    const char *runArg = "-Srun";
    char path[MAX_BUF_SIZE];
    char cmd[MAX_BUF_SIZE];
    const char *flag = "-f";
    const char *quote = "\"";
    DWORD platform = svcPlatform();


    if (configFile == NULL)
    {
        flag = "";
        quote = "";
        configFile = "";
    }

    /*
    ** Get the filename of the running executable. We need to leave
    ** space in the buffer for additional arguments, whitespace and
    ** quotes ...
    */

    if (GetModuleFileName(NULL, path,
                          MAX_BUF_SIZE - (strlen(configFile) + 20)) == 0)
    {
        message(0, 0, "can't get executable path");
        return EXIT_FAILURE;
    }

    /*
    ** Build the command string. If a config file was specified
    ** this will be of the form:
    **
    **  "program path" -f "config file" -Srun
    */

    sprintf(cmd, "\"%s\" %s %s%s%s %s", path,
            flag, quote, configFile, quote, runArg);

    switch (platform)
    {
    case VER_PLATFORM_WIN32_WINDOWS:
        return svcInstall9X(name, cmd);
        break;

    case VER_PLATFORM_WIN32_NT:
        return svcInstallNT(name, cmd);
        break;

    default:
        message(0, 0, "unsupported OS platform (type %d)", platform);
        break;
    }
    return EXIT_FAILURE;
}

int
svcInstallNT(char *name, char *cmd)
{
    SC_HANDLE svcHandle;
    SC_HANDLE scmHandle;


    /* Open the local SCM database */

    if ((scmHandle = OpenSCManager(NULL, NULL, SC_MANAGER_ALL_ACCESS)) == NULL)
    {
        message(0, 0, "can't open SCM database");
        return EXIT_FAILURE;
    }

    /* Now create the service entry */

    svcHandle = CreateService(scmHandle,    /* SCM database handle */
                              name,         /* Name of service */
                              name,         /* Display name */
                              SERVICE_ALL_ACCESS,       /* Desired access */
                              SERVICE_WIN32_OWN_PROCESS,/* Run in own process */
                              SERVICE_AUTO_START,       /* Start during boot */
                              SERVICE_ERROR_NORMAL,     /* Show message but continue boot */
                              cmd,          /* Program plus arguments */
                              NULL,         /* No load ordering group */
                              NULL,         /* No tag identifier */
                              NULL,         /* No dependencies */
                              NULL,         /* Use LocalSystem account */
                              NULL);        /* No password */
    CloseServiceHandle(scmHandle);

    if (svcHandle == NULL)
    {
        message(0, 0, "can't create service");
        return EXIT_FAILURE;
    }

    CloseServiceHandle(svcHandle);

    message(1, 0, "%s service installed to run '%s'", name, cmd);

    return EXIT_SUCCESS;
}

int
svcInstall9X(char *name, char *cmd)
{
    HKEY    runKey;


    /* Open RunServices registry key */

    if (RegCreateKey(HKEY_LOCAL_MACHINE,
                     "Software\\Microsoft\\Windows\\CurrentVersion\\RunServices",
                     &runKey) != ERROR_SUCCESS)
    {
        message(0, 0, "can't locate registry key");
        return EXIT_FAILURE;
    }

    /* Add value for this service */

    if (RegSetValueEx(runKey, name, 0, REG_SZ, cmd, strlen(cmd) + 1) != ERROR_SUCCESS)
    {
        RegCloseKey(runKey);
        message(0, 0, "failed to add value to registry");
        return EXIT_FAILURE;
    }

    RegCloseKey(runKey);

    message(1, 0, "%s service installed to run '%s'", name, cmd);

    return EXIT_SUCCESS;
}

int
svcRemove(char *name)
{
    DWORD platform = svcPlatform();


    switch (platform)
    {
    case VER_PLATFORM_WIN32_WINDOWS:
        return svcRemove9X(name);
        break;

    case VER_PLATFORM_WIN32_NT:
        return svcRemoveNT(name);
        break;

    default:
        message(0, 0, "unsupported OS platform (type %d)", platform);
        break;
    }

    return EXIT_FAILURE;
}

int
svcRemoveNT(char *name)
{
    SC_HANDLE   svcHandle;
    SC_HANDLE   scmHandle;
    SERVICE_STATUS status;
    int         retStatus = EXIT_SUCCESS;


    /* Open the SCM */

    if ((scmHandle = OpenSCManager(NULL, NULL, SC_MANAGER_ALL_ACCESS)) != NULL)
    {
        svcHandle = OpenService(scmHandle, name, SERVICE_ALL_ACCESS);

        if (svcHandle != NULL)
        {
            /* Try to stop the service */

            if (ControlService(svcHandle, SERVICE_CONTROL_STOP, &status))
            {
                while (QueryServiceStatus(svcHandle, &status))
                {
                    if (status.dwCurrentState == SERVICE_STOP_PENDING)
                    {
                        Sleep(1000);
                    }
                    else
                    {
                        break;
                    }
                }

                if (status.dwCurrentState != SERVICE_STOPPED)
                {
                    message(0, 0, "failed to stop the '%s' service", name);
                    retStatus = EXIT_FAILURE;
                }
            }

            /* Now remove the service from the SCM */

            if (DeleteService(svcHandle))
            {
                message(1, 0, "successfully removed the '%s' service", name);
            }
            else
            {
                message(0, 0, "failed to remove the '%s' service", name);
                retStatus = EXIT_FAILURE;
            }

            CloseServiceHandle(svcHandle);
        }
        else
        {
            message(0, 0, "can't find the '%s' service", name);
            retStatus = EXIT_FAILURE;
        }

        CloseServiceHandle(scmHandle);
    }
    else
    {
        message(0, 0, "can't contact Service Control Manager");
        retStatus = EXIT_FAILURE;
    }

    return retStatus;
}


int
svcRemove9X(char *name)
{
    HKEY runKey;
    int status = EXIT_SUCCESS;


    /* Locate the RunServices registry entry */

    if (RegOpenKey(HKEY_LOCAL_MACHINE, 
                   "Software\\Microsoft\\Windows\\CurrentVersion\\RunServices",
                   &runKey) != ERROR_SUCCESS)
    {
        message(0, 0, "can't find the RunServices registry key");
        status = EXIT_FAILURE;
    }
    else
    {
        /* Delete the key value for our service */

        if (RegDeleteValue(runKey, name) != ERROR_SUCCESS)
        {
            message(0, 0, "failed to delete registry RunServices entry for '%s'", name);
            status = EXIT_FAILURE;
        }
        else
        {
            message(1, 0, "successfully removed the '%s' service", name);
        }

        RegCloseKey(runKey);
    }

    return status;
}
